### PR TITLE
Add required versions to Terraform

### PR DESCRIPTION
**Pull request by austinkelleher**
@eXtremeX

### CODE REVIEW

**Review by aiwilliams — APPROVED** [review]
It seems to me this is an exceptional practice. How about specifying the major version of Terraform as well? I use `asdf` to manage my terraform versions, and an entry in `.tool-versions` to make my life easier. 😁

**Comment by austinkelleher** [comment]
@aiwilliams Agreed, but I think we should use the `required_version` constraint instead: https://www.terraform.io/docs/language/settings/index.html#specifying-a-required-terraform-version

**Comment by aiwilliams** [comment]
> @aiwilliams Agreed, but I think we should use the `required_version` constraint instead: https://www.terraform.io/docs/language/settings/index.html#specifying-a-required-terraform-version

Oh, yeah, totally meant that we should add the constraint to the terraform config. I should have said, "I also use `asdf` to manage my installed and selected terraform version."

**Review by ndowmon — DISMISSED** [review]
Interesting. Azure actually allows you to specify the version _on_ the [provider](https://github.com/JupiterOne/graph-azure/blob/master/terraform/main.tf#L1):

```
provider "azurerm" {
  version = "~>2.43.0"
  features {
    key_vault {
      recover_soft_deleted_key_vaults = true
      purge_soft_delete_on_destroy = true
    }
  }
}
```

But this is great to know.

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,8 @@
 node_modules/
 dist/
 .j1-integration/
-.env
+.env*
 .eslintcache
 tsconfig.tsbuildinfo
 coverage/
+main.remote.tf
```
Comment (austinkelleher) [inline .gitignore:8]: This file will be used by integration developers to target a specific Terraform Cloud workspace. Example:

```terraform
terraform {
  backend "remote" {
    organization = "my-tf-cloud-org"

    workspaces {
      name = "my-tf-org-workspace"
    }
  }
}
```

File: terraform/cloud_storage_buckets.tf [hunk 2]
```diff
@@ -10,60 +10,6 @@ resource "google_kms_crypto_key" "bucket_key" {
   depends_on = [google_kms_key_ring.bucket_key_ring]
 }
 
-resource "google_project_iam_member" "grant_google_storage_service_encrypt_decrypt" {
```
Comment (austinkelleher) [inline terraform/cloud_storage_buckets.tf:13]: Decided that none of these old resources were necessary anymore and were also causing various TF permission issues.

File: terraform/cloud_storage_buckets.tf [hunk 3]
```diff
@@ -10,60 +10,6 @@ resource "google_kms_crypto_key" "bucket_key" {
...
-  bucket = google_storage_bucket.public_storage_bucket.name
-  policy_data = data.google_iam_policy.public_storage_object_viewer.policy_data
-}
-################################################################################
-# End public bucket code
-################################################################################
-
 resource "google_storage_bucket" "sink-log-bucket" {
-  name = "my-unique-sink-logging-bucket"
+  name = "${var.project_id}-sink-logging-bucket"
```
Comment (austinkelleher) [inline terraform/cloud_storage_buckets.tf:14]: Needs to be globally unique

File: terraform/main.tf [hunk 4]
```diff
@@ -1,4 +1,6 @@
 terraform {
+  required_version = "~> 0.14"
```
Comment (austinkelleher) [inline terraform/main.tf:2]: 🥳

File: terraform/main.tf [hunk 5]
```diff
@@ -18,7 +20,7 @@ terraform {
 }
 
 provider "google" {
-  credentials = file(abspath(var.service_account_key_file_path))
+  credentials = var.service_account_key
```
Comment (austinkelleher) [inline terraform/main.tf:23]: Now inside of Terraform Cloud, but could still be specified locally if the user did not have a Terraform Cloud account.

**Review by aiwilliams — APPROVED** [review]

**Pull request merged**